torch/distributions/distribution.py

```diff
@@ -96,7 +96,15 @@ def _validate_log_prob_arg(self, value):
         """
         if not (torch.is_tensor(value) or isinstance(value, Variable)):
             raise ValueError('The value argument to log_prob must be a Tensor or Variable instance.')
-        batch_dim_start = len(value.size()) - len(self._batch_shape) - len(self._event_shape)
-        if value.size()[batch_dim_start:] != self._batch_shape + self._event_shape:
-            raise ValueError('The right-most size of value must match: {}.'.
-                             format(self._batch_shape + self._event_shape))
+
+        event_dim_start = len(value.size()) - len(self._event_shape)
+        if value.size()[event_dim_start:] != self._event_shape:
+            raise ValueError('The right-most size of value must match event_shape: {} vs {}.'.
+                             format(value.size(), self._event_shape))
+
+        actual_shape = value.size()
+        expected_shape = self._batch_shape + self._event_shape
+        for i, j in zip(reversed(actual_shape), reversed(expected_shape)):
```

Comment:
It may be cheaper to do
```
    if self._batch_shape and value.size()[:event_dim_start]:
        try:
            torch._C._infer_size(value.size()[:event_dim_start], self._batch_shape)
        except RuntimeError:
            raise ValueError('Value is not broadcastable with batch_shape+event_shape: {} vs {}.'.
                             format(actual_shape, expected_shape))
```

Just to make sure, this broadcasting will allow both upcasting of distribution parameters as well as upcasting of values:

e.g.
```
        p = Normal(Variable(torch.Tensor([[0.0], [1.0]])),
                   Variable(torch.Tensor([[1.0], [1.0]])))    # size (2, 1)
        x = Variable(torch.Tensor([0.0, 1.0]))    # size (2,)
        pdf = torch.exp(p.log_prob(x))     # size (2, 2)
```